### PR TITLE
Add support for nd-range kernels and local memory

### PR DESCRIPTION
**Pull request by fknorr**
With this patch, all `parallel_for` overloads allow specifying the work-group size by accepting an `nd_range` parameter. The work-group size is respected in the split, which receives a non-1 _granularity_ parameter. Kernels in these invocations will receive a `celerity::nd_item`.

`celerity::nd_item` and `celerity::group` provide equivalent functionality to their `sycl::` counterparts. Celerity exposes group-cooperative functions such as `celerity::reduce_over_group` as long as they are present in the SYCL implementation. We guarantee that at least `celerity::group_barrier` is present, which is a minimum requirement for any meaningful use of local memory.

Capturing a `celerity::local_accessor` allocates local memory to a kernel. Support for this is unfortunately limited to DPC++ (using an ugly hack for default-constructing the corresponding SYCL accessor) and a yet-to-be-released version of hipSYCL (cf. illuhad/hipSYCL#639). The `mat_mul` example has been updated to use local memory tiling when available.

To maintain our support for a wide variety of SYCL implementations and versions, the Celerity CMake project now detects SYCL features at configuration time, exposing three feature-detection macros (both in CMake as `ON`/`OFF` as well as in C++ defined to either `0` or `1`):

- `CELERITY_FEATURE_LOCAL_ACCESSOR` if Celerity can support local accessors with the given SYCL implementation
- `CELERITY_FEATURE_SIMPLE_SCALAR_REDUCTIONS` for the limited reduction support provided by DPC++
- `CELERITY_FEATURE_SCALAR_REDUCTIONS` for the full reduction support provided by illuhad/hipSYCL#578

### CODE REVIEW

**Review by psalz — CHANGES_REQUESTED** [review]
Thank you, a few minor notes; otherwise looks very good! Also as just discussed in person, please remove the relevant section from our "Issues & Limitations" documentation!

File: include/sycl_wrappers.h [hunk 1]
```diff
@@ -43,4 +43,19 @@ namespace property {
 
 inline const property::no_init no_init;
 
+// We do not expose memory_scope::system because Celerity does not expose USM or multi-GPU kernel launches
+enum class memory_scope { work_item, sub_group, work_group, device };
```
[the remark this answers is not in the document]
Comment (fknorr) [reply include/sycl_wrappers.h:52]: Done, with the exception of ComputeCpp which does not have this enum.

File: include/item.h [hunk 2]
```diff
@@ -56,4 +111,452 @@ class item {
...
+		return sycl_item.async_work_group_copy(dest, src, num_elements, dest_stride);
+	}
+
+	template <typename... DeviceEvents>
+	void wait_for(DeviceEvents... events) const {
+		sycl_item.wait_for(events...);
+	}
+
+  private:
+	// We capture SYCL `item` instead of `group` to provide celerity::group_barrier based on SYCL 1.2.1 nd_item.barrier()
```
[the remark this answers is not in the document]
Comment (fknorr) [reply include/item.h:180]: Done.

**Review by psalz — APPROVED** [review]

File: CMakeLists.txt [hunk 3]
```diff
@@ -216,12 +250,10 @@ if(CELERITY_BUILD_EXAMPLES)
...
   )
 
-  # CCPP currently does not implement reductions
-  if(CELERITY_SYCL_IMPL STREQUAL "DPC++" OR (
-    CELERITY_SYCL_IMPL STREQUAL "hipSYCL" AND CELERITY_HIPSYCL_SUPPORTS_REDUCTIONS))
+  if (CELERITY_FEATURE_SIMPLE_SCALAR_REDUCTIONS)
     add_subdirectory(examples/reduction)
     set_property(TARGET reduction PROPERTY FOLDER "examples")
-  endif()
+  endif ()
```
Comment (BlackMark29A) [inline CMakeLists.txt:256]: Formatting is inconsistent with the rest of the cmake file.

File: test/runtime_tests.cc [hunk 4]
```diff
@@ -2434,7 +2433,85 @@ namespace detail {
 #endif
 	}
 
-#endif // !WORKAROUND_COMPUTECPP && (!WORKAROUND_HIPSYCL || CELERITY_HIPSYCL_SUPPORTS_REDUCTIONS)
+#endif
+
+	TEST_CASE("handler::parallel_for accepts nd_range", "[handler]") {
+		distr_queue q;
+
+		CHECK_NOTHROW(q.submit([&](handler& cgh) { //
```
Comment (BlackMark29A) [inline test/runtime_tests.cc:2441]: Why the `//`?

**Review by BlackMark29A — APPROVED** [review]

**Pull request merged**